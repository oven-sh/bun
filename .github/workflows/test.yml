name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
