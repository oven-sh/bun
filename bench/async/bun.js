import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
