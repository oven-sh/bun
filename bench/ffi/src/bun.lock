{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
