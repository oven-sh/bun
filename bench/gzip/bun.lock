{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
