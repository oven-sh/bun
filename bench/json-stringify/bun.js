import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
