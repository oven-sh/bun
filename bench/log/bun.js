import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
