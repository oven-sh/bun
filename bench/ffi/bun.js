import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
