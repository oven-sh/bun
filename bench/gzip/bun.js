import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
