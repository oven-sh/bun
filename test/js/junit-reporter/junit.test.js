import { it, describe, expect } from "bun:test";
import { file, spawn } from "bun";
import { tempDirWithFiles, bunEnv, bunExe } from "harness";

const xml2js = require("xml2js");

describe("junit reporter", () => {
  for (let withCIEnvironmentVariables of [false, true]) {
    it(`should generate valid junit xml for passing tests ${withCIEnvironmentVariables ? "with CI environment variables" : ""}`, async () => {
      const tmpDir = tempDirWithFiles("junit", {
        "package.json": "{}",
        "passing.test.js": `

        it("should pass", () => {
          expect(1 + 1).toBe(2);
        });

        it("second test", () => {
          expect(1 + 1).toBe(2);
        });

        it("failing test", () => {
          expect(1 + 1).toBe(3);
        });

        it.skip("skipped test", () => {
          expect(1 + 1).toBe(2);
        });

        it.todo("todo test");

        describe("nested describe", () => {
          it("should pass inside nested describe", () => {
            expect(1 + 1).toBe(2);
          });

          it("should fail inside nested describe", () => {
            expect(1 + 1).toBe(3);
          });
        });
      `,

        "test-2.test.js": `

        it("should pass", () => {
          expect(1 + 1).toBe(2);
        });

        it("failing test", () => {
          expect(1 + 1).toBe(3);
        });

        describe("nested describe", () => {
          it("should pass inside nested describe", () => {
            expect(1 + 1).toBe(2);
          });

          it("should fail inside nested describe", () => {
            expect(1 + 1).toBe(3);
          });
        });
      `,
      });

      let env = bunEnv;

      if (withCIEnvironmentVariables) {
        env = {
          ...env,
          CI_JOB_URL: "https://ci.example.com/123",
          CI_COMMIT_SHA: "1234567890",
        };
      }

      const junitPath = `${tmpDir}/junit.xml`;
      const proc = spawn([bunExe(), "test", "--reporter=junit", "--reporter-outfile", junitPath], {
        cwd: tmpDir,
        env,
        stdout: "inherit",
        "stderr": "inherit",
      });
      await proc.exited;
      console.log(junitPath);

      expect(proc.exitCode).toBe(1);
      const xmlContent = await file(junitPath).text();

      // Parse XML to verify structure
      const result = await new Promise((resolve, reject) => {
        xml2js.parseString(xmlContent, (err, result) => {
          if (err) reject(err);
          else resolve(result);
        });
      });

      /**
     * ------ Vitest ------
     * <?xml version="1.0" encoding="UTF-8" ?>
     * <testsuites name="vitest tests" tests="11" failures="4" errors="0" time="0.176">
     *     <testsuite name="passing.test.js" timestamp="2024-11-18T09:21:11.933Z" hostname="Jarreds-MacBook-Pro.local" tests="7" failures="2" errors="0" skipped="2" time="0.005116208">
     *         <testcase classname="passing.test.js" name="should pass" time="0.000657541">
     *         </testcase>
     *         <testcase classname="passing.test.js" name="second test" time="0.000071875">
     *         </testcase>
     *         <testcase classname="passing.test.js" name="failing test" time="0.003308209">
     *             <failure message="expected 2 to be 3 // Object.is equality" type="AssertionError">
     * AssertionError: expected 2 to be 3 // Object.is equality
     * 
     * - Expected
     * + Received
     * 
     * - 3
     * + 2
     * 
     * ❯ passing.test.js:12:25
     *             </failure>
     *         </testcase>
        <testcase classname="passing.test.js" name="skipped test" time="0">
     *             <skipped/>
     *         </testcase>
     *         <testcase classname="passing.test.js" name="todo test" time="0">
     *             <skipped/>
     *         </testcase>
     *          <testcase classname="passing.test.js" name="nested describe &gt; should pass inside nested describe" time="0.000130042">
     *         </testcase>
     *         <testcase classname="passing.test.js" name="nested describe &gt; should fail inside nested describe" time="0.000403125">
     *             <failure message="expected 2 to be 3 // Object.is equality" type="AssertionError">
     * AssertionError: expected 2 to be 3 // Object.is equality
     * 
     * - Expected
     * + Received
     * 
     * - 3
     * + 2
     * 
     * ❯ passing.test.js:27:27
     *             </failure>
     *         </testcase>
    </testsuite>
     *     <testsuite name="test-2.test.js" timestamp="2024-11-18T09:21:11.936Z" hostname="Jarreds-MacBook-Pro.local" tests="4" failures="2" errors="0" skipped="0" time="0.005188916">
     *         <testcase classname="test-2.test.js" name="should pass" time="0.000642541">
     *         </testcase>
     *         <testcase classname="test-2.test.js" name="failing test" time="0.003380708">
     *             <failure message="expected 2 to be 3 // Object.is equality" type="AssertionError">
     * AssertionError: expected 2 to be 3 // Object.is equality
     * 
     * - Expected
     * + Received
     * 
     * - 3
     * + 2
     * 
     * ❯ test-2.test.js:8:25
     *             </failure>
     *         </testcase>
     *         <testcase classname="test-2.test.js" name="nested describe &gt; should pass inside nested describe" time="0.000140541">
     *         </testcase>
     *         <testcase classname="test-2.test.js" name="nested describe &gt; should fail inside nested describe" time="0.000306">
     *             <failure message="expected 2 to be 3 // Object.is equality" type="AssertionError">
     * AssertionError: expected 2 to be 3 // Object.is equality
     * 
     * - Expected
     * + Received
     * 
     * - 3
     * + 2
     * 
     * ❯ test-2.test.js:17:27
     *             </failure>
     *         </testcase>
     *     </testsuite>
     * </testsuites>
     */

      expect(result.testsuites).toBeDefined();
      expect(result.testsuites.testsuite).toBeDefined();

      let firstSuite = result.testsuites.testsuite[0];
      let secondSuite = result.testsuites.testsuite[1];

      if (firstSuite.$.name === "passing.test.js") {
        [firstSuite, secondSuite] = [secondSuite, firstSuite];
      }

      expect(firstSuite.testcase).toHaveLength(4);
      expect(firstSuite.testcase[0].$.name).toBe("should pass inside nested describe");
      expect(firstSuite.$.name).toBe("test-2.test.js");
      expect(firstSuite.$.tests).toBe("4");
      expect(firstSuite.$.failures).toBe("2");
      expect(firstSuite.$.skipped).toBe("0");
      expect(parseFloat(firstSuite.$.time)).toBeGreaterThanOrEqual(0.0);

      expect(secondSuite.testcase).toHaveLength(7);
      expect(secondSuite.testcase[0].$.name).toBe("should pass inside nested describe");
      expect(secondSuite.$.name).toBe("passing.test.js");
      expect(secondSuite.$.tests).toBe("7");
      expect(secondSuite.$.failures).toBe("2");
      expect(secondSuite.$.skipped).toBe("2");
      expect(parseFloat(secondSuite.$.time)).toBeGreaterThanOrEqual(0.0);

      expect(result.testsuites.$.tests).toBe("11");
      expect(result.testsuites.$.failures).toBe("4");
      expect(result.testsuites.$.skipped).toBe("2");
      expect(parseFloat(result.testsuites.$.time)).toBeGreaterThanOrEqual(0.0);

      if (withCIEnvironmentVariables) {
        // "properties": [
        //   {
        //     "property": [
        //       {
        //         "$": {
        //           "name": "ci",
        //           "value": "https://ci.example.com/123"
        //         }
        //       },
        //       {
        //         "$": {
        //           "name": "commit",
        //           "value": "1234567890"
        //         }
        //       }
        //     ]
        //   }
        // ],
        expect(firstSuite.properties).toHaveLength(1);
        expect(firstSuite.properties[0].property).toHaveLength(2);
        expect(firstSuite.properties[0].property[0].$.name).toBe("ci");
        expect(firstSuite.properties[0].property[0].$.value).toBe("https://ci.example.com/123");
        expect(firstSuite.properties[0].property[1].$.name).toBe("commit");
        expect(firstSuite.properties[0].property[1].$.value).toBe("1234567890");
      }
    });
  }
});
