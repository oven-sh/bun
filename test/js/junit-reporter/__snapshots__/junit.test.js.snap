// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`junit reporter more scenarios 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="bun test" tests="22" assertions="16" failures="1" skipped="6">
  <testsuite name="comprehensive.test.js" file="comprehensive.test.js" tests="22" assertions="16" failures="1" skipped="6">
    <testsuite name="comprehensive test suite" file="comprehensive.test.js" line="4" tests="22" assertions="16" failures="1" skipped="6">
      <testsuite name="division suite 10 / 5" file="comprehensive.test.js" line="8" tests="1" assertions="1" failures="0" skipped="0">
        <testcase name="should divide correctly" classname="division suite 10 / 5 &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="9" assertions="1" />
      </testsuite>
      <testsuite name="division suite 20 / 10" file="comprehensive.test.js" line="8" tests="1" assertions="1" failures="0" skipped="0">
        <testcase name="should divide correctly" classname="division suite 20 / 10 &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="9" assertions="1" />
      </testsuite>
      <testsuite name="conditional describe that runs" file="comprehensive.test.js" line="14" tests="1" assertions="1" failures="0" skipped="0">
        <testcase name="nested test in conditional describe" classname="conditional describe that runs &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="15" assertions="1" />
      </testsuite>
      <testsuite name="conditional describe that skips" file="comprehensive.test.js" line="20" tests="1" assertions="0" failures="0" skipped="1">
        <testcase name="nested test that gets skipped" classname="conditional describe that skips &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="21" assertions="0">
          <skipped />
        </testcase>
      </testsuite>
      <testcase name="basic passing test" classname="comprehensive test suite" file="comprehensive.test.js" line="26" assertions="1" />
      <testcase name="basic failing test" classname="comprehensive test suite" file="comprehensive.test.js" line="30" assertions="1">
        <failure type="AssertionError" />
      </testcase>
      <testcase name="basic skipped test" classname="comprehensive test suite" file="comprehensive.test.js" line="34" assertions="0">
        <skipped />
      </testcase>
      <testcase name="basic todo test" classname="comprehensive test suite" file="comprehensive.test.js" line="38" assertions="0">
        <skipped message="TODO" />
      </testcase>
      <testcase name="addition 1 + 2 = 3" classname="comprehensive test suite" file="comprehensive.test.js" line="44" assertions="1" />
      <testcase name="addition 2 + 3 = 5" classname="comprehensive test suite" file="comprehensive.test.js" line="44" assertions="1" />
      <testcase name="addition 4 + 5 = 9" classname="comprehensive test suite" file="comprehensive.test.js" line="44" assertions="1" />
      <testcase name="string concat hello + world = helloworld" classname="comprehensive test suite" file="comprehensive.test.js" line="51" assertions="1" />
      <testcase name="string concat foo + bar = foobar" classname="comprehensive test suite" file="comprehensive.test.js" line="51" assertions="1" />
      <testcase name="conditional test that runs" classname="comprehensive test suite" file="comprehensive.test.js" line="55" assertions="1" />
      <testcase name="conditional test that skips" classname="comprehensive test suite" file="comprehensive.test.js" line="59" assertions="0">
        <skipped />
      </testcase>
      <testcase name="skip if true" classname="comprehensive test suite" file="comprehensive.test.js" line="63" assertions="0">
        <skipped />
      </testcase>
      <testcase name="skip if false" classname="comprehensive test suite" file="comprehensive.test.js" line="67" assertions="1" />
      <testcase name="todo if true" classname="comprehensive test suite" file="comprehensive.test.js" line="71" assertions="0">
        <skipped message="TODO" />
      </testcase>
      <testcase name="todo if false" classname="comprehensive test suite" file="comprehensive.test.js" line="73" assertions="1" />
      <testcase name="test marked as failing" classname="comprehensive test suite" file="comprehensive.test.js" line="77" assertions="1" />
      <testcase name="should match this test" classname="comprehensive test suite" file="comprehensive.test.js" line="81" assertions="1" />
      <testcase name="should not be matched by filter" classname="comprehensive test suite" file="comprehensive.test.js" line="85" assertions="1" />
    </testsuite>
  </testsuite>
</testsuites>
"
`;

exports[`junit reporter more scenarios 2`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="bun test" tests="22" assertions="1" failures="0" skipped="21">
  <testsuite name="comprehensive.test.js" file="comprehensive.test.js" tests="22" assertions="1" failures="0" skipped="21">
    <testsuite name="comprehensive test suite" file="comprehensive.test.js" line="4" tests="22" assertions="1" failures="0" skipped="21">
      <testsuite name="division suite 10 / 5" file="comprehensive.test.js" line="8" tests="1" assertions="0" failures="0" skipped="1">
        <testcase name="should divide correctly" classname="division suite 10 / 5 &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="9" assertions="0">
          <skipped />
        </testcase>
      </testsuite>
      <testsuite name="division suite 20 / 10" file="comprehensive.test.js" line="8" tests="1" assertions="0" failures="0" skipped="1">
        <testcase name="should divide correctly" classname="division suite 20 / 10 &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="9" assertions="0">
          <skipped />
        </testcase>
      </testsuite>
      <testsuite name="conditional describe that runs" file="comprehensive.test.js" line="14" tests="1" assertions="0" failures="0" skipped="1">
        <testcase name="nested test in conditional describe" classname="conditional describe that runs &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="15" assertions="0">
          <skipped />
        </testcase>
      </testsuite>
      <testsuite name="conditional describe that skips" file="comprehensive.test.js" line="20" tests="1" assertions="0" failures="0" skipped="1">
        <testcase name="nested test that gets skipped" classname="conditional describe that skips &amp;gt; comprehensive test suite" file="comprehensive.test.js" line="21" assertions="0">
          <skipped />
        </testcase>
      </testsuite>
      <testcase name="basic passing test" classname="comprehensive test suite" file="comprehensive.test.js" line="26" assertions="0">
        <skipped />
      </testcase>
      <testcase name="basic failing test" classname="comprehensive test suite" file="comprehensive.test.js" line="30" assertions="0">
        <skipped />
      </testcase>
      <testcase name="basic skipped test" classname="comprehensive test suite" file="comprehensive.test.js" line="34" assertions="0">
        <skipped />
      </testcase>
      <testcase name="basic todo test" classname="comprehensive test suite" file="comprehensive.test.js" line="38" assertions="0">
        <skipped />
      </testcase>
      <testcase name="addition 1 + 2 = 3" classname="comprehensive test suite" file="comprehensive.test.js" line="44" assertions="0">
        <skipped />
      </testcase>
      <testcase name="addition 2 + 3 = 5" classname="comprehensive test suite" file="comprehensive.test.js" line="44" assertions="0">
        <skipped />
      </testcase>
      <testcase name="addition 4 + 5 = 9" classname="comprehensive test suite" file="comprehensive.test.js" line="44" assertions="0">
        <skipped />
      </testcase>
      <testcase name="string concat hello + world = helloworld" classname="comprehensive test suite" file="comprehensive.test.js" line="51" assertions="0">
        <skipped />
      </testcase>
      <testcase name="string concat foo + bar = foobar" classname="comprehensive test suite" file="comprehensive.test.js" line="51" assertions="0">
        <skipped />
      </testcase>
      <testcase name="conditional test that runs" classname="comprehensive test suite" file="comprehensive.test.js" line="55" assertions="0">
        <skipped />
      </testcase>
      <testcase name="conditional test that skips" classname="comprehensive test suite" file="comprehensive.test.js" line="59" assertions="0">
        <skipped />
      </testcase>
      <testcase name="skip if true" classname="comprehensive test suite" file="comprehensive.test.js" line="63" assertions="0">
        <skipped />
      </testcase>
      <testcase name="skip if false" classname="comprehensive test suite" file="comprehensive.test.js" line="67" assertions="0">
        <skipped />
      </testcase>
      <testcase name="todo if true" classname="comprehensive test suite" file="comprehensive.test.js" line="71" assertions="0">
        <skipped />
      </testcase>
      <testcase name="todo if false" classname="comprehensive test suite" file="comprehensive.test.js" line="73" assertions="0">
        <skipped />
      </testcase>
      <testcase name="test marked as failing" classname="comprehensive test suite" file="comprehensive.test.js" line="77" assertions="0">
        <skipped />
      </testcase>
      <testcase name="should match this test" classname="comprehensive test suite" file="comprehensive.test.js" line="81" assertions="1" />
      <testcase name="should not be matched by filter" classname="comprehensive test suite" file="comprehensive.test.js" line="85" assertions="0">
        <skipped />
      </testcase>
    </testsuite>
  </testsuite>
</testsuites>
"
`;
